src/util/deferred.ts

```diff
@@ -1,29 +1,31 @@
 export class Deferred<T> {
   readonly #promise: Promise<T>
 
-  #resolve?: (value: T | PromiseLike<T>) => void
   #reject?: (reason?: any) => void
+  #resolve?: (value: T | PromiseLike<T>) => void
 
   constructor() {
     this.#promise = new Promise<T>((resolve, reject) => {
-      this.#reject = reject
-      this.#resolve = resolve
+      this.#reject = (reason?: any) => {
+        reject(reason)
+        this.#reject = this.#resolve = undefined
```

Comment:
Should you clear #reject and #resolve before the `reject(reason)` call? I don't really remember how promises worked in this situation, but can `reject(reason)` trigger code that calls `this.reject` or `this.resolve`?